File: Dashboard/src/views/Projects/taskPickupRules.js
export const PICKUP_FIELDS = [
  { id: "author", label: "Author", hint: "Logins or stable user IDs, e.g. alice, bob" },
  { id: "assignee", label: "Assignee", hint: "Tracker logins/IDs, or local actor IDs" },
  { id: "queue", label: "Tracker queue", hint: "Queue keys, e.g. CORE, MOBILE" },
  { id: "issue_type", label: "Tracker issue type", hint: "Type keys, e.g. bug, task" },
  { id: "priority", label: "Priority", hint: "high, medium, low, or a Tracker priority key" },
  { id: "external_status", label: "Tracker status", hint: "Status keys, e.g. open, inProgress" },
  { id: "tag", label: "Tag", hint: "Tag values, e.g. ios, manual" },
  { id: "title", label: "Title", hint: "Title text" },
  { id: "description", label: "Description", hint: "Description text" },
  { id: "source", label: "Source", hint: "startrek, github, or local" },
  { id: "kind", label: "Local task kind", hint: "planning, execution, bugfix" }
];

export const PICKUP_OPERATORS = [
  { id: "one_of", label: "is one of" },
  { id: "not_one_of", label: "is not one of" },
  { id: "contains", label: "contains text" },
  { id: "is_set", label: "is present" },
  { id: "is_not_set", label: "is missing" }
];

export function clonePickupRules(rules) {
  return {
    matchMode: rules?.matchMode || "all",
    conditions: Array.isArray(rules?.conditions) ? rules.conditions.map((condition) => ({
      ...condition,
      values: Array.isArray(condition.values) ? [...condition.values] : []
    })) : []
  };
}

export function pickupRulesError(rules) {
  if (!["all", "any"].includes(rules.matchMode)) return "Choose how conditions are combined.";
  if (rules.conditions.length > 50) return "Use at most 50 conditions.";
  for (const [index, condition] of rules.conditions.entries()) {
    if (!PICKUP_FIELDS.some((field) => field.id === condition.field) ||
        !PICKUP_OPERATORS.some((operator) => operator.id === condition.operation)) {
      return `Condition ${index + 1} is unsupported. Remove it or choose a supported field and operator.`;
    }
    if (condition.operation === "contains" && !["title", "description"].includes(condition.field)) {
      return `Condition ${index + 1}: text matching is available for title and description.`;
    }
    if (!["is_set", "is_not_set"].includes(condition.operation) && !condition.values.some((value) => String(value).trim())) {
      return `Add a value to condition ${index + 1}.`;
    }
  }
  return "";
}

export function pickupIdentityOptions(tasks, field) {
  if (!["author", "assignee"].includes(field)) return [];
  const options = new Map();
  for (const task of tasks) {
    const metadata = task.externalMetadata;
    const identity = field === "author" ? metadata?.externalCreator : metadata?.externalAssigneeIdentity;
    const value = String(metadata?.providerId
      ? identity?.login || identity?.id || ""
      : (field === "author" ? task.createdBy : task.actorId) || "").trim();
    if (!value) continue;
    const display = identity?.displayName;
    options.set(value.toLowerCase(), { id: value, label: display && display !== value ? `${display} (${value})` : value });
  }
  return [...options.values()].sort((left, right) => left.label.localeCompare(right.label));
}

File: Sources/Protocols/TaskPickupRules.swift
import Foundation

public enum TaskPickupField: String, Codable, Sendable, CaseIterable {
    case author, assignee, queue, priority, tag, title, description, source, kind
    case issueType = "issue_type"
    case externalStatus = "external_status"
}

public enum TaskPickupOperator: String, Codable, Sendable, CaseIterable {
    case oneOf = "one_of"
    case notOneOf = "not_one_of"
    case contains
    case isSet = "is_set"
    case isNotSet = "is_not_set"
}

public struct TaskPickupCondition: Codable, Sendable, Equatable, Identifiable {
    public var id: String
    public var field: TaskPickupField
    public var operation: TaskPickupOperator
    public var values: [String]

    public init(id: String = UUID().uuidString, field: TaskPickupField, operation: TaskPickupOperator = .oneOf, values: [String] = []) {
        self.id = id
        self.field = field
        self.operation = operation
        self.values = values
    }

    public var isValid: Bool {
        if operation == .contains && field != .title && field != .description { return false }
        return operation == .isSet || operation == .isNotSet || !normalizedPickupValues(values).isEmpty
    }

    public func matches(_ task: ProjectTask) -> Bool {
        guard isValid else { return false }
        let actual = normalizedPickupValues(task.pickupValues(for: field))
        if operation == .isSet { return !actual.isEmpty }
        if operation == .isNotSet { return actual.isEmpty }
        // Tags are a known list, including when that list is empty.
        if field == .tag && operation == .notOneOf && actual.isEmpty { return true }
        // Unknown source metadata must not pass a negative identity condition.
        guard !actual.isEmpty else { return false }
        let expected = normalizedPickupValues(values)
        switch operation {
        case .oneOf: return !Set(actual).isDisjoint(with: expected)
        case .notOneOf: return Set(actual).isDisjoint(with: expected)
        case .contains: return actual.contains { value in expected.contains { value.contains($0) } }
        case .isSet, .isNotSet: return false
        }
    }
}

public struct ProjectTaskPickupRules: Codable, Sendable, Equatable {
    public enum MatchMode: String, Codable, Sendable { case all, any }
    public var matchMode: MatchMode
    public var conditions: [TaskPickupCondition]

    public init(matchMode: MatchMode = .all, conditions: [TaskPickupCondition] = []) {
        self.matchMode = matchMode
        self.conditions = conditions
    }

    public var isValid: Bool {
        conditions.count <= 50 && conditions.allSatisfy(\.isValid)
    }

    public func matches(_ task: ProjectTask) -> Bool {
        guard isValid else { return false }
        guard !conditions.isEmpty else { return true }
        switch matchMode {
        case .all: return conditions.allSatisfy { $0.matches(task) }
        case .any: return conditions.contains { $0.matches(task) }
        }
    }
}

private func normalizedPickupValues(_ values: [String]) -> [String] {
    values.map { $0.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() }.filter { !$0.isEmpty }
}

extension ProjectTask {
    public func pickupValues(for field: TaskPickupField) -> [String] {
        let metadata = externalMetadata
        switch field {
        case .author:
            if metadata?.providerId != nil {
                return [metadata?.externalCreator?.id, metadata?.externalCreator?.login].compactMap { $0 }
            }
            return [createdBy].compactMap { $0 }
        case .assignee:
            if metadata?.providerId != nil {
                return [metadata?.externalAssigneeIdentity?.id, metadata?.externalAssigneeIdentity?.login].compactMap { $0 }
            }
            return [actorId].compactMap { $0 }
        case .queue: return [metadata?.externalQueue].compactMap { $0 }
        case .issueType: return [metadata?.externalIssueType].compactMap { $0 }
        case .priority: return [priority, metadata?.externalPriorityKey].compactMap { $0 }
        case .externalStatus: return [metadata?.externalStatus?.key].compactMap { $0 }
        case .tag: return tags
        case .title: return [title]
        case .description: return [description]
        case .source: return [metadata?.providerId ?? "local"]
        case .kind: return [kind?.rawValue].compactMap { $0 }
        }
    }
}
